src/goto-harness/goto_harness_parse_options.cpp

```diff
@@ -6,47 +6,183 @@ Author: Diffblue Ltd.
 
 \******************************************************************/
 
+#include "goto_harness_parse_options.h"
+
 #include <cstddef>
-#include <iostream>
+#include <set>
 #include <string>
+#include <utility>
 
+#include <goto-programs/goto_convert.h>
+#include <goto-programs/goto_model.h>
+#include <goto-programs/read_goto_binary.h>
+#include <goto-programs/write_goto_binary.h>
+#include <util/config.h>
+#include <util/exception_utils.h>
 #include <util/exit_codes.h>
+#include <util/invariant.h>
 #include <util/version.h>
 
-#include "goto_harness_parse_options.h"
+#include "function_call_harness_generator.h"
+#include "goto_harness_generator_factory.h"
+
+// The basic idea is that this module is handling the following
+// sequence of events:
+// 1. Initialise a goto-model by parsing an input (goto) binary
+// 2. Initialise the harness generator (with some config) that will handle
+//    the mutation of the goto-model. The generator should create a new
+//    function that can be called by `cbmc --function`. The generated function
+//    should implement the behaviour of the harness (What exactly this means
+//    depends on the configuration)
+// 3. Write the end result of that process to the output binary
 
 int goto_harness_parse_optionst::doit()
 {
   if(cmdline.isset("version"))
   {
-    std::cout << CBMC_VERSION << '\n';
+    logger.status() << CBMC_VERSION << '\n';
     return CPROVER_EXIT_SUCCESS;
   }
 
-  help();
-  return CPROVER_EXIT_USAGE_ERROR;
+  auto got_harness_config = handle_common_options();
+  auto factory = make_factory();
+
+  auto factory_options = collect_generate_factory_options();
+
+  // Initialise harness generator
+  auto harness_generator =
+    factory.factory(got_harness_config.harness_type, factory_options);
+  CHECK_RETURN(harness_generator != nullptr);
+
+  // TODO: What does this actually do? Do we need to configure the command
+  //   line options for this to work?
+  config.set(cmdline);
```

Comment:
This just sets up the defaults (and would interpret options such as `--32`). I'm not sure that's really required here, but it's probably a good idea to have a `config` in a sane state.

Reply:
We had a problem with binaries generated by `goto_cc`. The problem is that while we are trying to call `config.set_from_symbol_table(goto_model.symbol_table);` from a binary that was compiled with `goto-cc`, we hit an `UNREACHABLE` assertion in `configt::ansi_ct::set_arch_spec_x86_64()`. It has `flavourt::NONE`.

Reply:
Hmm, yes, so you need to set the platform-specific defaults via this call.